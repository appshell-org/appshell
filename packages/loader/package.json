{
  "name": "@appshell/loader",
  "version": "0.4.0",
  "description": "Utility for loading Appshell components",
  "main": "dist/main.js",
  "types": "dist/loader/src/index.d.ts",
  "repository": "https://github.com/navaris/appshell.git",
  "author": "Robert Hamilton <rh@navaris.com>",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "scripts": {
    "build": "webpack --mode production && npm run build:types",
    "build:types": "tsc --emitDeclarationOnly",
    "clean": "rm -rf dist",
    "prebuild": "npm run clean",
    "start": "webpack --watch --mode development && npm run build:types",
    "stats": "webpack --json > stats.json"
  },
  "keywords": [
    "module federation",
    "micro-frontends",
    "appshell"
  ],
  "license": "MIT"
}
