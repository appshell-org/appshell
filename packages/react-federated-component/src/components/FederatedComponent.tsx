/* eslint-disable react/jsx-props-no-spreading */
import remoteLoader from '@appshell/loader';
import React, { ComponentType, ReactElement, ReactNode, useEffect, useState } from 'react';
import useManifest from '../hooks/useManifest';
import LoadingError from './LoadingError';

export type ExtendedProps = Record<string, unknown>;

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
export type FederatedComponentProps<TProps extends ExtendedProps = ExtendedProps> = {
  remote: string;
  fallback?: ReactNode;
} & TProps;

const FederatedComponent = <TProps extends ExtendedProps>({
  remote,
  fallback,
  ...rest
}: FederatedComponentProps<TProps>): ReactElement<TProps> => {
  const manifest = useManifest();
  const [element, setElement] = useState<ReactElement>();

  useEffect(() => {
    let active = false;
    const loadComponent = remoteLoader(manifest);

    async function load() {
      try {
        active = true;
        setElement(undefined);
        const Component = await loadComponent<ComponentType>(remote);
        if (!Component) {
          return;
        }
        active = false;
        setElement(React.createElement(Component, rest));
      } catch (err) {
        setElement(<LoadingError remote={remote} reason={`${err}`} />);
      }
    }

    if (!active) {
      load();
    }

    return () => {
      active = false;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // eslint-disable-next-line no-console
  console.debug(`rendering FederatedComponent[${remote}], loading=${!element}`);
  // eslint-disable-next-line react/jsx-no-useless-fragment
  return <>{element || fallback}</>;
};

FederatedComponent.defaultProps = {
  fallback: undefined,
};

export default FederatedComponent;
