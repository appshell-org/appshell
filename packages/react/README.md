<div align="center">
  <a href="https://github.com/navaris/appshell">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://github.com/navaris/appshell/blob/main/assets/branding/appshell-logo-white_2x.png">
      <img alt="appshell" src="https://github.com/navaris/appshell/blob/main/assets/branding/appshell-logo_2x.png">
    </picture>
  </a>
</div>

[![Appshell CI](https://github.com/navaris/appshell/actions/workflows/pipeline.yml/badge.svg)](https://github.com/navaris/appshell/actions/workflows/pipeline.yml)

# @appshell/react

React utilites for building micro-frontends with Webpack Module federation and Appshell.

Working examples can be found [here](https://github.com/navaris/appshell/tree/main/examples).

## Getting Started

To begin, you'll need to install `@appshell/react`:

```console
npm install @appshell/react --save-dev
```

or

```console
yarn add -D @appshell/react
```

or

```console
pnpm add -D @appshell/react
```

## FederatedComponent

React component that dynamically loads federated components.

```tsx
import { FederatedComponent, ManifestProvider } from '@appshell/react';

<App>
  <ManifestProvider manifest={manifest}>
    <FederatedComponent remote="PingModule/Ping">
    <FederatedComponent remote="PongModule/Pong">
  </ManifestProvider>
</App>
```

## useRegistry

For access to the registry index.

```tsx
import { RegistryProvider, useRegistry } from '@appshell/react';

const MyComponent = () => {
  const index = useRegistry();

  ...
}

<RegistryProvider index={index}>
  <MyComponent />
</RegistryProvider>
```

## useMetadata

For access to the registry metadata.

```tsx
import { MetadataProvider, useMetadata } from '@appshell/react';

const MyComponent = () => {
  const metadata = useMetadata();

  ...
}

<MetadataProvider metadata={metadata}>
  <MyComponent />
</MetadataProvider>
```

## useManifest

For access to the manifest.

```tsx
import { ManifestProvider, useManifest } from '@appshell/react';

const MyComponent = () => {
  const manifest = useManifest();

  ...
}

<ManifestProvider manifest={manifest}>
  <MyComponent />
</ManifestProvider>
```

**Where does the registry come from?**

Currently, the registry is a shared directory or mounted volume. It does support calling an http endpoint, but it has not yet been implemented.

> See [@appshell/cli](https://www.npmjs.com/package/@appshell/cli)

Each micro-frontend registers its manifest with the registry

```bash
appshell register --manifest dist/appshell.manifest.json
```
