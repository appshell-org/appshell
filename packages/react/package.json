{
  "name": "@appshell/react",
  "version": "0.3.1-alpha.0",
  "description": "React utilities for building micro-frontends with Appshell and Module Federation",
  "main": "dist/main.js",
  "types": "dist/types/react/src/index.d.ts",
  "repository": "https://github.com/navaris/appshell.git",
  "author": "Robert Hamilton <rh@navaris.com>",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "scripts": {
    "build": "webpack --mode production && npm run build:types",
    "build:types": "tsc --emitDeclarationOnly",
    "clean": "rm -rf dist",
    "start": "webpack --watch --mode development --stats-error-details",
    "stats": "webpack --json > stats.json"
  },
  "keywords": [
    "module federation",
    "micro-frontends",
    "appshell"
  ],
  "license": "MIT",
  "peerDependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  }
}
