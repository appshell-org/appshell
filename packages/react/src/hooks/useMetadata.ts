import { Metadata } from '@appshell/config';
import React from 'react';
// eslint-disable-next-line import/no-named-as-default
import MetadataContext from '../contexts/MetadataContext';

export default (): Metadata => React.useContext<Metadata>(MetadataContext);
