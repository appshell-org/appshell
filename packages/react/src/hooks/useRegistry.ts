import { AppshellIndex } from '@appshell/config';
import React from 'react';
// eslint-disable-next-line import/no-named-as-default
import RegistryContext from '../contexts/RegistryContext';

export default (): AppshellIndex => React.useContext<AppshellIndex>(RegistryContext);
