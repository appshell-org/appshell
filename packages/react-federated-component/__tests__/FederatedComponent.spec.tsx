/** @jest-environment jsdom */
import * as remoteLoader from '@appshell/loader';
import '@testing-library/jest-dom';
import { act, render, screen } from '@testing-library/react';
import React from 'react';
import FederatedComponent from '../src/components/FederatedComponent';
import * as useManifest from '../src/hooks/useManifest';
import manifest from './fixtures/Manifest';
import TestComponent from './fixtures/TestComponent';

jest.mock('../src/hooks/useManifest');

const TestFallback = () => <div>loading...</div>;

describe('FederatedComponent', () => {
  it('should match snapshot', async () => {
    jest.spyOn(useManifest, 'default').mockReturnValue(manifest);
    jest.spyOn(remoteLoader, 'default').mockReturnValueOnce(async () => TestComponent);

    const { container, findByText } = await act(() =>
      render(<FederatedComponent remote="TestModule/TestComponent" />),
    );
    const view = await findByText(/test component/i);

    expect(view).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render fallback while resource is pending', async () => {
    jest.spyOn(useManifest, 'default').mockReturnValue(manifest);
    jest.spyOn(remoteLoader, 'default').mockReturnValueOnce(() => null);

    await act(() =>
      render(<FederatedComponent remote="TestModule/TestComponent" fallback={<TestFallback />} />),
    );

    await expect(screen.getByText(/loading/i)).toBeInTheDocument();
  });

  it('should render error when resource fails', async () => {
    jest.spyOn(useManifest, 'default').mockReturnValue(manifest);
    jest
      .spyOn(remoteLoader, 'default')
      .mockImplementationOnce(() => new Error('Failed to get resource'));

    await act(() =>
      render(<FederatedComponent remote="TestModule/TestComponent" fallback={<TestFallback />} />),
    );

    await expect(screen.getByText(/Error loading federated component/i)).toBeInTheDocument();
  });
});
