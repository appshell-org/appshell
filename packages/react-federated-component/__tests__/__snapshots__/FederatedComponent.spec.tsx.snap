// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FederatedComponent should match snapshot 1`] = `
<div>
  <div>
    test component
  </div>
</div>
`;
